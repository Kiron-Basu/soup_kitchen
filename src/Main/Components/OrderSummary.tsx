import React from "react";
import { IDiscount } from "../Interfaces/IDiscount";
import { poundify } from "../Utilities/InvoiceUtilities";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import { Typography } from "@material-ui/core";
import "../Styles/OrderSummary.scss";
import { Button, Divider } from "@material-ui/core"; // import all like this
import MoneyOffIcon from "@material-ui/icons/MoneyOff";

interface IOrderSummary {
  preDiscountTotal: number;
  discounts: ReadonlyArray<IDiscount>;
}

export const OrderSummary: React.FC<IOrderSummary> = ({
  preDiscountTotal,
  discounts,
}) => {
  return (
    <List className="orderSummary">
      {preDiscountTotal <= 0 ? <h5>Your cart is empty!</h5> : null}
      {renderDiscounts()}
      {renderSubTotal()}
      {renderTotal()}
      <Button
        variant="contained"
        color="secondary"
        size="large"
        disabled={preDiscountTotal <= 0}
      >
        Order
      </Button>
    </List>
  );

  function renderSubTotal() {
    return preDiscountTotal > 0 ? (
      <Typography variant="h6">
        <Divider />
        <div className="orderSummary__subtotal">
          Subtotal: {poundify(preDiscountTotal)}
        </div>
      </Typography>
    ) : null;
  }

  function renderTotal() {
    const grandTotal: number =
      preDiscountTotal -
      discounts.reduce((acc, val) => {
        return (acc += val.discount);
      }, 0);
    return grandTotal > 0 ? (
      <Typography variant="h5" gutterBottom={true}>
        <Divider />
        <div className="orderSummary__total">Total: {poundify(grandTotal)}</div>
        <Divider />
      </Typography>
    ) : null;
  }

  function renderDiscounts(): Array<JSX.Element | null> | JSX.Element | null {
    if (preDiscountTotal <= 0) {
      return null;
    }

    return discounts.some((discount) => discount.discount > 0) ? (
      discounts.map((discount) =>
        discount.discount > 0 ? (
          <ListItem>
            <MoneyOffIcon fontSize="small" />
            {`${discount.type}`} : {poundify(discount.discount)}
          </ListItem>
        ) : null
      )
    ) : (
      <ListItem>No discounts applied!</ListItem>
    );
  }
};
