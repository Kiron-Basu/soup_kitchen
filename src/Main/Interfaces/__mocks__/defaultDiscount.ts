import { IDiscount } from "../IDiscount";

export const defaultDiscount: IDiscount = {
  type: "Bread discount",
  discount: 100,
};
