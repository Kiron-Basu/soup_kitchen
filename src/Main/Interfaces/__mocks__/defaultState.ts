export const defaultState = {
  bread: {
    price: 110,
    quantity: 0,
    subtotal: 0,
  },
  butter: {
    price: 120,
    quantity: 0,
    subtotal: 0,
  },
  cheese: {
    price: 90,
    quantity: 0,
    subtotal: 0,
  },
  milk: {
    price: 50,
    quantity: 0,
    subtotal: 0,
  },
  soup: {
    price: 60,
    quantity: 0,
    subtotal: 0,
  },
};
