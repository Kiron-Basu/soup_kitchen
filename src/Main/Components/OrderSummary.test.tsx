import React from "react";
import { 
  ShallowWrapper,
  shallow 
} from "enzyme";

import { OrderSummary } from "./OrderSummary";
import { defaultDiscount } from "../Interfaces/__mocks__/defaultDiscount";

describe("OrderSummary tests", () => {
  test("Renders OrderSummary correctly", () => {
    const component: ShallowWrapper = shallow(
      <OrderSummary discounts={[defaultDiscount]} preDiscountTotal={200} />
    );

    expect(component).toMatchSnapshot();
  });
});
