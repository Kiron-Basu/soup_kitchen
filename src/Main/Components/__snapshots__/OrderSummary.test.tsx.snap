// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary tests Renders OrderSummary correctly 1`] = `
<WithStyles(ForwardRef(List))
  className="orderSummary"
>
  <WithStyles(ForwardRef(ListItem))
    key="0discount"
  >
    <Memo(ForwardRef(MoneyOffIcon))
      fontSize="small"
    />
    Bread discount
     : 
    £1.00
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Typography))
    variant="h6"
  >
    <WithStyles(ForwardRef(Divider)) />
    <div
      className="orderSummary__subtotal"
    >
      Subtotal: 
      £2.00
    </div>
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="h5"
  >
    <WithStyles(ForwardRef(Divider)) />
    <div
      className="orderSummary__total"
    >
      Total: 
      £1.00
    </div>
    <WithStyles(ForwardRef(Divider)) />
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Button))
    color="secondary"
    disabled={false}
    size="large"
    variant="contained"
  >
    Order
  </WithStyles(ForwardRef(Button))>
</WithStyles(ForwardRef(List))>
`;
